### PR TITLE
Refactor Bundle into Graph and Bundle

### PR DESCRIPTION
**Pull request by guybedford**
The goal of this work is that by allowing Bundle to represent multiple separate chunks on the same graph, we can create automated code-splitting workflows.

### CODE REVIEW

**Comment by guybedford** [comment]
This is now ready for review.

I will be posting two further PRs that build on top of this:
1. Dynamic import inlining by default
2. A new code-splitting output for Rollup

These should hopefully be coming this week with some luck, so it would be good to start getting this in.

It will make review of (1) and (2) above much simpler as well.

File: test/form/samples/internal-conflict-resolution/_expected/amd.js [hunk 1]
```diff
@@ -1,15 +1,15 @@
 define(function () { 'use strict';
 
-	var bar$1 = 42;
+	var bar = 42;
```
Comment (guybedford) [inline test/form/samples/internal-conflict-resolution/_expected/amd.js]: Note these changes are a result of running deshadowing in execution order (due to it running on bundle.orderedModules over graph.modules). There are no other repercussions of this approach though.

Comment (guybedford) [reply test/form/samples/internal-conflict-resolution/_expected/amd.js]: Reverted - please ignore.

**Comment by lukastaegert** [comment]
Hi, I see you have been really busy! I’m currently with family and only have a phone available but I’ll try to sneak in a review.

**Comment by guybedford** [comment]
Thanks @lukastaegert. Note this doesn't make any code changes apart from splitting up the functions and classes and moving things around, so the diff looks worse than it really is. Enjoy the family holiday there, this can wait till the new year fine, no pressure :)

**Comment by guybedford** [comment]
@lukastaegert thanks for the feedback. Would you be happy for me to merge then?

I split out part of the logic into a separate private `mark` function, which did some simplification here. I think good architecture will always have code reductions as a side effect, so to always focus on improving the architecture should become a process itself of moving more logic to its natural place of reuse. In a good simple architecture, all high-level operations should break down into simpler low level ones in a way that is intuitive, so it's about keeping an eye to always simplifying the picture as much as possible (but no simpler :P). </abstract coding generalisations>.

For example, there's a whole bunch of logic around external module pruning in the algorithm. This seems like something that could be moved into the external module class itself, and then run over each external module as a higher level operation. Handling these "decentralizations" of logic, may require tracking parent information on variables and exports - then we would be able to determine each parent module doing the importing of a given external export from that external export's class itself, and have more analysis that side. Just some armchair thoughts, without thinking about it too deeply though, I may just be talking rubbish again.

**Comment by lukastaegert** [comment]
Will need to think a little about what you said last but otherwise yes, please do the merge!

**Comment by guybedford** [comment]
Sure, thanks, merging now! Just got some code splitting working here... looks like I should have that PR up to schedule :)

**Pull request merged**